hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/TestLatencyTracker.java

```diff
@@ -0,0 +1,479 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.regex.Pattern;
+
+/**
+ * Test the latency tracker for abfs
+ *
+ */
+public final class TestLatencyTracker {
+  private static final Logger LOG = LoggerFactory.getLogger(TestLatencyTracker.class);
+  private static final int TEST_AGGREGATE_LATENCY = 42;
+  private final String filesystemName = "bogusFilesystemName";
+  private final String accountName = "bogusAccountName";
+  private final URL url;
+
+  public TestLatencyTracker() throws Exception {
+    this.url = new URL("http", "www.microsoft.com", "/bogusFile");
+  }
+
+  @Test
+  public void verifyDisablingOfTracker() throws Exception {
+    // verify that disabling of the tracker works
+    LatencyTracker latencyTracker = new LatencyTracker(accountName, filesystemName, false);
+
+    String latencyDetails = latencyTracker.getClientLatency();
+    Assert.assertNull("LatencyTracker should be empty", latencyDetails);
+
+    latencyTracker.recordClientLatency(Instant.now(), "disablingCaller", "disablingCallee", true,
+            new AbfsHttpOperation(url, "GET", new ArrayList<AbfsHttpHeader>()));
+
+    latencyDetails = latencyTracker.getClientLatency();
+    Assert.assertNull("LatencyTracker should return no record", latencyDetails);
+  }
+
+  @Test
+  public void verifyTrackingForSingletonLatencyRecords() throws Exception {
+    // verify that tracking for singleton latency records works as expected
+    final int numTasks = 100;
+    LatencyTracker latencyTracker = new LatencyTracker(accountName, filesystemName, true);
+
+    String latencyDetails = latencyTracker.getClientLatency();
+    Assert.assertNull("LatencyTracker should be empty", latencyDetails);
+
+    ExecutorService executorService = Executors.newCachedThreadPool();
```

Comment:
shut this down at the end of the test

Reply:
Shut down is now in the @afterTest method.